crypto/crypto_aes_aesni.c

```diff
@@ -177,22 +177,19 @@ crypto_aes_key_expand_aesni(const uint8_t * key, size_t len)
 }
 
 /**
- * crypto_aes_encrypt_block_aesni(in, out, key):
- * Using the expanded AES key ${key}, encrypt the block ${in} and write the
- * resulting ciphertext to ${out}.  ${in} and ${out} can overlap.  This
- * implementation uses x86 AESNI instructions, and should only be used if
- * CPUSUPPORT_X86_AESNI is defined and cpusupport_x86_aesni() returns nonzero.
+ * crypto_aes_encrypt_block_aesni_direct(in, key):
```

Comment:
I'm not sold on the name `_direct()`.  I considered `_sse2()` but that looked worse.

Reply:
I would be inclined to go with `crypto_aes_encrypt_block_aesni_m128i`.

Reply:
Thanks.  I'm reconsidering the whole approach to crypto_aesctr, though, so it's probably best to hold off on reviewing them more.  It looks like we might be able to get 90% of the speed boost with a mess less invasive code change.  If that works out, then I'll submit that as a separate PR, then look at the fancier stuff later (if it even makes sense).

Reply:
Ok.  FWIW, I'm inclined to put this into a different header file which only gets included if CPUSUPPORT_X86_AESNI rather than having it internally decide whether to include the header needed for __m128i.

Also, I would seriously consider having the "aesctr aesni" special case only used for encrypting complete blocks at a time.  For the pre-block and post-block encryption we can take the regular slower path...

Reply:
That's tempting, but unfortunately having any function call inside the `_stream` function immediately drops performance by ~33% (397 to 308 MB/s for buffer size of 1024).  Using SSE2 instructions to xor the buffers creeps us back up to 410 MB/s so technically it's still an overall gain, but nothing near as nice as I was hoping for.  I tested the performance penalty by adding a `_wholeblock()` function to `crypto_aesctr.c` itself, then adding the non-standard `__attribute__((noinline))` to it.

At the moment, I'm leaning towards having some unfortunate duplication of code between `crypto_aesctr.c` and `crypto_aesctr_aesni.c`.  I like the idea of a `_wholeblock` function in `crypto_aesctr_aesni.c`, becaues that will let us use a `crypto_aes_encrypt_block_aesni_m128i()` with a temporary `__m128i` variable, without worrying about memory alignment of `stream->buf`.

(I expect that it will still be useful to do some memory alignment of `stream`, but it would be cleaner to handle that as a separate question from the mere existance of `crypto_aesctr_aesni.c` itself.)